### PR TITLE
Add NCS specific changes

### PR DESCRIPTION
**Pull request by simensrostad**
Add NCS specific changes that was previously added on-top of an internal copy of cJSON in sdk-nrf.

**Note**
Upstream cJSON is build with C89 so the changes added in https://github.com/nrfconnect/sdk-cjson/commit/098fc8394057e0f6b55a03fe3c2639e5e6d31294 will most likely not be added upstream due to ```snprintf```.

As for https://github.com/nrfconnect/sdk-cjson/commit/842de7544dbe5203a1803e8aeacc25083c16e28d I don't know if this is something that could be committed upstream or not. The change seems trivial for most applications, but if it is an optimization that accounts for all users of cJSON it makes sense to add it.

AFAIK there is nothing that prohibits https://github.com/nrfconnect/sdk-cjson/commit/b2f80f37542285e8e0d2f2a2b3c90369e3d13faf from being added upstream as it only adds new APIs on pre-existing functionality to cJSON. So I marked it ```toup```.

### CODE REVIEW

**Review by sigvartmh — APPROVED** [review]
Seems to contain the `strcpy`to`strncpy` and `sprintf` and `snprintf`. So they are now using more safe functions. Thanks Simen :)

File: cJSON.h [hunk 1]
```diff
@@ -255,7 +266,7 @@ CJSON_PUBLIC(cJSON *) cJSON_Duplicate(const cJSON *item, cJSON_bool recurse);
 CJSON_PUBLIC(cJSON_bool) cJSON_Compare(const cJSON * const a, const cJSON * const b, const cJSON_bool case_sensitive);
 
 /* Minify a strings, remove blank characters(such as ' ', '\t', '\r', '\n') from strings.
- * The input pointer json cannot point to a read-only address area, such as a string constant, 
+ * The input pointer json cannot point to a read-only address area, such as a string constant,
```
Comment (mbolivar-nordic) [inline cJSON.h]: Please remove the extraneous whitespace change. That seems unrelated to this commit's purpose.

File: cJSON.c [hunk 2]
```diff
@@ -92,19 +92,19 @@ CJSON_PUBLIC(const char *) cJSON_GetErrorPtr(void)
     return (const char*) (global_error.json + global_error.position);
 }
 
-CJSON_PUBLIC(char *) cJSON_GetStringValue(const cJSON * const item) 
+CJSON_PUBLIC(char *) cJSON_GetStringValue(const cJSON * const item)
```
Comment (mbolivar-nordic) [inline cJSON.c]: Please don't mix large-scale whitespace and functional changes like this. It makes things harder to review and adds diff from upstream, which is a source of merge conflicts when synchronizing. Can you go through the commits in this PR and remove the intermixing of whitespace and functional changes within a single commit?

If you're going to fix whitespace nits, I suggest doing it upstream first, since they are not bugs, but style issues, and we want to keep it simple when it comes to avoiding merge conflicts when syncing with upstream later. That means minimizing diff from upstream.

File: cJSON.c [hunk 3]
```diff
@@ -558,22 +558,22 @@ static cJSON_bool print_number(const cJSON * const item, printbuffer * const out
     /* This checks for NaN and Infinity */
     if (isnan(d) || isinf(d))
     {
-        length = sprintf((char*)number_buffer, "null");
+        length = snprintf((char*)number_buffer, sizeof(number_buffer), "null");
```
Comment (mbolivar-nordic) [inline cJSON.c:561]: What is the justification for having this be a `noup` patch? Did upstream reject it? If so, what is wrong with their justification; if not, can we send this upstream and keep it here as a `fromlist` patch instead?

File: cJSON.c [hunk 4]
```diff
@@ -479,7 +479,7 @@ static unsigned char* ensure(printbuffer * const p, size_t needed)
     }
     else
     {
-        newsize = needed * 2;
+        newsize = needed + 1;
```
Comment (mbolivar-nordic) [inline cJSON.c:482]: Same question here; any reason not to send this upstream and see what they say, instead of carrying a `noup` patch?

File: cJSON.c [hunk 5]
```diff
@@ -2057,6 +2057,124 @@ CJSON_PUBLIC(cJSON_bool) cJSON_AddItemToObjectCS(cJSON *object, const char *stri
     return add_item_to_object(object, string, item, &global_hooks, true);
 }
 
+CJSON_PUBLIC(void) cJSON_AddItemReferenceToObjectCS(cJSON *object, const char *string, cJSON *item)
```
Comment (mbolivar-nordic) [inline cJSON.c:2060]: We generally try to avoid `toup` patches unless it's an emergency and there really is no time to submit upstream first.

Can you please:

1. submit this patch upstream (without any `[nrf ...]` tag, of course)
2. change this commit to `fromlist`
3. add a link to the upstream PR in the commit log of the downstream `fromlist` patch?

If upstream rejects it, we can carry a `noup`. If upstream is slow to respond, we can carry a `fromlist`. But let's try to get this upstreamed, please.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Thanks very much for putting this together.

I have a few requests but this looks very good already.

File: cJSON.c (hunk 3 above)
Comment (mbolivar-nordic) [reply cJSON.c:561]: Never mind, I see the note in the PR description that explains why upstream probably won't want this.

Can you please put it into the commit log, so we don't lose this information?

File: cJSON.c (hunk 4 above)
Comment (mbolivar-nordic) [reply cJSON.c:482]: Likewise I see the note in the PR description now. I still think we should submit it upstream and see what they say.

If upstream has a configuration mechanism for fine-tuning the library behavior, perhaps this can leverage that to make this settable in NCS.

File: cJSON.h (hunk 1 above)
Comment (simensrostad) [reply cJSON.h]: Jepp. Done.

File: cJSON.c (hunk 3 above)
Comment (simensrostad) [reply cJSON.c:561]: Updated the commit log explaining why this patch cannot go upstream.

File: cJSON.c (hunk 2 above)
Comment (simensrostad) [reply cJSON.c]: :+1:

**Comment by simensrostad** [comment]
@mbolivar-nordic Added https://github.com/nrfconnect/sdk-cjson/pull/1/commits/7509bd2bda1463a14b3b6a2c6913ecdefb5e862f and https://github.com/nrfconnect/sdk-cjson/pull/1/commits/61407f339b0c1fc7c6374aa3b180c0398d1bb85a upstream and changed the tags to [nrf fromlist]. I also resolved the whitespace issues that you commented on.

File: cJSON.c (hunk 2 above)
Comment (simensrostad) [reply cJSON.c]: I see that I might need to look this one over again.

Comment (simensrostad) [reply cJSON.c]: Should be OK.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by tejlmand — CHANGES_REQUESTED** [review]
I believe this should should be made into a Zephyr module by adding `zephyr/module.yml`:
```
build:
    cmake-ext: True
    kconfig-ext: True

```

**Comment by simensrostad** [comment]
> I believe this should should be made into a Zephyr module by adding `zephyr/module.yml`:
> 
> ```
> build:
>     cmake-ext: True
>     kconfig-ext: True
> ```

Added ```zephyr/module.yml``` to root folder.

File: cJSON.c (hunk 5 above)
Comment (plskeggs) [reply cJSON.c:2060]: I missed this partner PR to https://github.com/nrfconnect/sdk-nrf/pull/4409 entirely.  Sorry about that.

Yes, I'll submit my changes upstream when I am less busy with the big push to get predicted GPS into NCS 1.6.0.  Hope more delay is acceptable.

Comment (plskeggs) [reply cJSON.c:2060]: Doh!  I stand corrected -- @simensrostad already did that.  Thanks!

**Comment by plskeggs** [comment]
LGTM

**Review by tejlmand — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
Sorry about this, but, I had to force push into master after merging this. This repository was not configured for rebase and merge, so clicking 'merge' created a merge commit, which we don't want. I force pushed just these changes. Please update the sdk-nrf PR with the final SHA which was merged into the main branch, 50bdddd.

**Comment by simensrostad** [comment]
> Sorry about this, but, I had to force push into master after merging this. This repository was not configured for rebase and merge, so clicking 'merge' created a merge commit, which we don't want. I force pushed just these changes. Please update the sdk-nrf PR with the final SHA which was merged into the main branch, [50bdddd](https://github.com/nrfconnect/sdk-cjson/commit/50bddddf7fb45e4ce87bc3619df06db55d181856).

No worries. The commit hash should now be up to date.

**Comment by bas-archembedded** [comment]
@mbolivar-nordic @simensrostad Can you please have a look at https://github.com/nrfconnect/sdk-cjson/pull/3?
